src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/CoverageLogic.kt

```diff
@@ -82,13 +120,65 @@ class RepositoriesCoverageLogic(private val modellingGroupRepository: ModellingG
                 )
             }.asSequence()
         }.toList()
+
+        val setsWithMissingRows = expectedRowLookup.missingRows()
+        if (setsWithMissingRows.any())
+        {
+            throw MissingRowsError(rowErrorMessage(setsWithMissingRows))
+        }
         touchstoneRepository.saveCoverageForTouchstone(touchstoneVersionId, records)
     }
 
-    constructor(repositories: Repositories) : this(repositories.modellingGroup,
-            repositories.responsibilities,
-            repositories.touchstone,
-            repositories.scenario)
+    private fun validate(vaccine: String,
+                         row: CoverageIngestionRow,
+                         countries: List<String>,
+                         expectedRowLookup: CoverageRowLookup): CoverageRowLookup
+    {
+        if (!expectedGAVICoverageYears.contains(row.year))
+        {
+            throw BadRequest("Unexpected year: ${row.year}")
+        }
+        if (row.activityType == ActivityType.ROUTINE)
+        {
+            val countryLookup = expectedRowLookup[vaccine] ?: generateCountryLookup(countries)
+            val year = row.year.toShort()
+            // note that although this country validation only happens for routine coverage rows,
+            // the foreign key constraint will catch invalid countries for campaign data
```

Comment:
But that just checks it's a country in the table, could still be a non-GAVI country? Again, I'm not really sure what the rules are here, just want to clarify!

Reply:
Correct. And this is a discrepancy between campaign and routine that I've left in for purely convenience reasons. Sorry I thought I had written a comment on the PR to this effect but it was on the PR that I closed!

Katy said that it didn't matter if they gave us a superset of countries (when we ship out the burden estimates we only use countries in the burden_estimate_country_expectation table anyway) but it would actually complicate the logic here to explicitly allow non-GAVI73 countries so in fact it won't accept any extra countries for routine data. For campaign data it will accept extra valid countries.

Reply:
I agree the discrepancy between routine and campaign is slightly weird here though.